data/build_redial.py

```diff
@@ -0,0 +1,131 @@
+# limitations under the License.
+"""Preprocessing Scripts for the Redial Dataset."""
+import json
+from tqdm import tqdm
+import numpy as np
+import re
+import os
+
+def prepare_redial(data_dir):
+    """Processes raw redial data in data_dir and saves the results in data folder."""
+    print("--Loading Redial Dataset--")
+    train = read_jsonl(os.path.join(data_dir, "rd-train-data.jsonl"))
+    test = read_jsonl(os.path.join(data_dir,"rd-test-data.jsonl"))
+
+
+    print("--Replacing Movie IDs--")
+    for dialogue in tqdm(train):
+        replace_ids(dialogue)
+    for dialogue in tqdm(test):
+        replace_ids(dialogue)
+
+    print("-Formatting For Training--")
+    train_formatted = separate_responses(train)
+    test_formatted = separate_responses(test)
+
+    write_jsonl("./redial/rd-train-formatted.jsonl", train_formatted)
+    write_jsonl("./redial/rd-test-formatted.jsonl", test_formatted)
+
+def length_summary(data_dir):
+    """prints a five number summary of the lengths of redial input/outputs."""
+    print("--Loading Dataset For Summary--")
+    train_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+    test_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+
+    def len_function(key):
+        return lambda x: len(x[key].split())
+
+    lengths = {"train_inputs" : list(map(len_function("conversation"), train_data)),
+                "train_targets" : list(map(len_function("response"), train_data)),
+                "test_inputs" : list(map(len_function("conversation"), test_data)),
+                "test_targets" : list(map(len_function("response"), test_data))}
+
+    for name, length_array in lengths.items():
+        print(name.upper() + ": ")
+        quartile_summary(length_array)
+
+#Helper Functions
+
+def read_jsonl(filename):
+    """Reads a jsonl file and returns as array of dicts."""
+    with open(filename, 'r') as json_file:
+        json_list = list(json_file)
+    data = []
+    for json_str in tqdm(json_list):
+        data.append(json.loads(json_str))
+    return data
+
+def write_jsonl(filename, arr):
+    """Writes array to jsonl."""
+    with open(filename, 'w') as f:
+        for line in arr:
+            f.write(json.dumps(line) + "\n")
+    
+def replace_ids(dialogue):
+    """Replaces movie ids in one redial dialogue with their corresponding movie titles."""
+    movie_titles = dialogue["movieMentions"]
+    for message in dialogue["messages"]:
+        text = message["text"]
+        replaced = []
+        for word in text.split():
+            if word[0] == "@" and re.sub('\\D', '', word) in movie_titles:
+                replaced.append( "@ " + movie_titles[re.sub('\\D', '', word)] + " @")
+            else:
+                replaced.append(word)
+        message["text"] = " ".join(replaced)
+
+def separate_responses(dataset):
+    """Creates a dataset of {"previous conversation" : "recommender response"} dictionaries
```

Comment:
Document the additional formatting (user/assistant prefixes) and also any additional assumptions you've made in setting up the training data format this way.

Reply:
added to the comment